### ISSUE TITLE
Type Props for actions and tasks

### ISSUE THREAD

**Issue by evantahler** [actionhero/actionhero#1400]
I'll bet we can use `inputs` to learn the types of `params` for actions and tasks...

**Comment by evantahler** [comment]
This is blocked by https://github.com/actionhero/actionhero/issues/1408

**Comment by evantahler** [comment]
The path might involve building a pseudo-class in the [ActionProcessor](https://github.com/actionhero/actionhero/blob/master/src/classes/actionProcessor.ts) that extends the types of the Action, perhaps an "Intersection type" (more her: https://www.typescriptlang.org/docs/handbook/advanced-types.html)

**Comment by papb** [comment]
Is it acceptable to expect [`as const`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions) being used when setting `this.inputs`?

```ts
this.inputs = {
  /* ... things here that I will think of and suggest soon ... */
} as const;
```

Or does it have to work without `as const`?

(if `as const` can be expected, it will be much easier to do this)

**Comment by evantahler** [comment]
I think const will work OK! `Action.inputs` are static and shouldn't change over the lifecycle of the application.

**Comment by evantahler** [comment]
Ping! @papb is there anything I can help with?  I'm really excited for typed Action inputs!

**Comment by papb** [comment]
Pong! Sorry for the delay, I have been quite busy, but hopefully next week I will be able to help you. Sorry!!

**Comment by evantahler** [comment]
https://github.com/sindresorhus/type-fest/blob/master/source/entries.d.ts looks relevant to our interests!

**Comment by krlicmuhamed** [comment]
How do I declare a Action Template variable `authenticate` in my example. 
```
export default class Status extends Action {
  constructor() {
    super();
    this.name = "status";
    this.description = "I will return some basic information about the API";
    this.outputExample = {
      id: "192.168.2.11",
      actionheroVersion: "9.4.1",
      uptime: 10469,
    };
    this.authenticate = true; // throws an error even though It is defined in an initializer
    // throws (src/actions/status.ts:26:10 - error TS2339: Property 'authenticate' does not exist on type 'Status'.)
  }
```
Initializer middleware code contains data.actionTemplate.authenticate ie.
```
        // Middleware code
        if (data.actionTemplate.authenticate && api.jwtauth.enabled[data.connection.type]) {
          tokenRequired = true;
        }
```
Thanks

**Comment by evantahler** [comment]
@krlicmuhamed in your case, you want to make a new class that extends `Action` and adds the new properties:

```ts
import { Action } from "actionhero";

export abstract class AuthenticatedAction extends Action {
  permission: {
    authenticate: boolean;
  };

  constructor() {
    super();

    this.middleware = ["authenticated-action"];
  }
}
```

And then use your new class rather than Action itself in your Actions:

```ts
export default class Status extends AuthenticatedAction {
  constructor() {
    super();
    this.name = "status";
    this.description = "I will return some basic information about the API";
    this.outputExample = {
      id: "192.168.2.11",
      actionheroVersion: "9.4.1",
      uptime: 10469,
    };
    this.authenticate = true;
  }
```


We do this in Grouparoo: 
* New Class: https://github.com/grouparoo/grouparoo/blob/master/core/src/classes/authenticatedAction.ts
* Usage: https://github.com/grouparoo/grouparoo/blob/master/core/src/actions/teamMembers.ts

**Comment by krlicmuhamed** [comment]
@evantahler hi, I tried this approach but I get `alert: Failed to load C:/Projects/pratibaApi/dist/actions/authenticatedAction.js stacktrace=Error: name is required for this action` for the abstract class AuthenticatedAction when I try to `npm run start`
Here's full stacktrace
```
2021-01-04T17:16:33.133Z - alert: Failed to load C:/Projects/pratibaApi/dist/actions/authenticatedAction.js stacktrace=Error: name is required for this action
    at AuthenticatedAction.validate (C:\Projects\pratibaApi\node_modules\actionhero\dist\classes\action.js:51:19)
    at Object.index_1.api.actions.loadFile (C:\Projects\pratibaApi\node_modules\actionhero\dist\initializers\actions.js:37:34)
    at async Actions.initialize (C:\Projects\pratibaApi\node_modules\actionhero\dist\initializers\actions.js:68:17)
    at async Object.initializeFunction (C:\Projects\pratibaApi\node_modules\actionhero\dist\classes\process.js:101:29)
    at async Object.asyncWaterfall (C:\Projects\pratibaApi\node_modules\actionhero\dist\modules\utils\asyncWaterfall.js:47:23)
    at async Process.initialize (C:\Projects\pratibaApi\node_modules\actionhero\dist\classes\process.js:170:13)
    at async Process.start (C:\Projects\pratibaApi\node_modules\actionhero\dist\classes\process.js:179:13)
    at async main (C:\Projects\pratibaApi\dist\server.js:13:5)
2021-01-04T17:16:33.152Z - emerg: Exception occurred in initializer `actions` during load 
2021-01-04T17:16:33.153Z - emerg: Error with initializer step: "initialize"
2021-01-04T17:16:33.154Z - emerg: Error: name is required for this action
    at AuthenticatedAction.validate (C:\Projects\pratibaApi\node_modules\actionhero\dist\classes\action.js:51:19)
    at Object.index_1.api.actions.loadFile (C:\Projects\pratibaApi\node_modules\actionhero\dist\initializers\actions.js:37:34)
    at async Actions.initialize (C:\Projects\pratibaApi\node_modules\actionhero\dist\initializers\actions.js:68:17)
    at async Object.initializeFunction (C:\Projects\pratibaApi\node_modules\actionhero\dist\classes\process.js:101:29)
    at async Object.asyncWaterfall (C:\Projects\pratibaApi\node_modules\actionhero\dist\modules\utils\asyncWaterfall.js:47:23)
    at async Process.initialize (C:\Projects\pratibaApi\node_modules\actionhero\dist\classes\process.js:170:13)
    at async Process.start (C:\Projects\pratibaApi\node_modules\actionhero\dist\classes\process.js:179:13)
    at async main (C:\Projects\pratibaApi\dist\server.js:13:5)
```

**Comment by evantahler** [comment]
@krlicmuhamed Can you share example code to reproduce the error?

**Comment by krlicmuhamed** [comment]
> @krlicmuhamed Can you share example code to reproduce the error?

Here you go https://github.com/krlicmuhamed/apiauth 
```
$ npm run build && npm run start 

> my_actionhero_project@0.1.0 build
> tsc --sourceMap false --declaration


> my_actionhero_project@0.1.0 start
> node ./dist/server.js

2021-01-05T15:25:54.820Z - alert: Failed to load C:/Projects/APIAUTH/dist/actions/classes/authenticatedAction.js stacktrace=Error: name is required for this action
    at AuthenticatedAction.validate (C:\Projects\APIAUTH\node_modules\actionhero\dist\classes\action.js:51:19)
    at Object.index_1.api.actions.loadFile (C:\Projects\APIAUTH\node_modules\actionhero\dist\initializers\actions.js:37:34)
    at processTicksAndRejections (node:internal/process/task_queues:93:5)
    at async Actions.initialize (C:\Projects\APIAUTH\node_modules\actionhero\dist\initializers\actions.js:68:17)
    at async Object.initializeFunction (C:\Projects\APIAUTH\node_modules\actionhero\dist\classes\process.js:101:29)
    at async Object.asyncWaterfall (C:\Projects\APIAUTH\node_modules\actionhero\dist\modules\utils\asyncWaterfall.js:47:23)
    at async Process.initialize (C:\Projects\APIAUTH\node_modules\actionhero\dist\classes\process.js:170:13)
    at async Process.start (C:\Projects\APIAUTH\node_modules\actionhero\dist\classes\process.js:179:13)
    at async main (C:\Projects\APIAUTH\dist\server.js:13:5)
2021-01-05T15:25:54.865Z - emerg: Exception occurred in initializer `actions` during load 
2021-01-05T15:25:54.867Z - emerg: Error with initializer step: "initialize"
2021-01-05T15:25:54.869Z - emerg: Error: name is required for this action
    at AuthenticatedAction.validate (C:\Projects\APIAUTH\node_modules\actionhero\dist\classes\action.js:51:19)
    at Object.index_1.api.actions.loadFile (C:\Projects\APIAUTH\node_modules\actionhero\dist\initializers\actions.js:37:34)
    at processTicksAndRejections (node:internal/process/task_queues:93:5)
    at async Actions.initialize (C:\Projects\APIAUTH\node_modules\actionhero\dist\initializers\actions.js:68:17)
    at async Object.initializeFunction (C:\Projects\APIAUTH\node_modules\actionhero\dist\classes\process.js:101:29)
    at async Object.asyncWaterfall (C:\Projects\APIAUTH\node_modules\actionhero\dist\modules\utils\asyncWaterfall.js:47:23)
    at async Process.initialize (C:\Projects\APIAUTH\node_modules\actionhero\dist\classes\process.js:170:13)
    at async Process.start (C:\Projects\APIAUTH\node_modules\actionhero\dist\classes\process.js:179:13)
    at async main (C:\Projects\APIAUTH\dist\server.js:13:5)
2021-01-05T15:25:54.904Z - crit: Cannot shut down actionhero, not running 
npm ERR! code 1
npm ERR! path C:\Projects\APIAUTH
npm ERR! command failed
npm ERR! command C:\Windows\system32\cmd.exe /d /s /c node ./dist/server.js

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Muhamed\AppData\Local\npm-cache\_logs\2021-01-05T15_25_55_980Z-debug.log
```

**Comment by evantahler** [comment]
Ah ha! Move that class out of the ‘actions’ folder. We look through that folder and try to load anything that extends Action... but that class isn’t a real action.  Can you open a new issue about this please?
